### ISSUE TITLE
Support for macOS 10.14 dark mode

### ISSUE THREAD

**Issue by singularitti** [xi-editor/xi-mac#276]
Will you consider support the macOS 10.14 Mojave's dark mode?
Expected features:
1. Dark title bar: just as what they do for Terminal.app:
<img width="639" alt="titlebar" src="https://user-images.githubusercontent.com/25192197/46174523-49366b00-c277-11e8-9522-e1e7197046b4.png">
2. Allow user to choose from dark and light titlebars
3. Dynamic title bar: The titlebar's color will change from light to dark when the sun is set (This feature may be too much so I do not really expect it to be implemented now)

**Comment by cmyr** [comment]
yep, definitely would be nice to do, good to have an issue for this.

**Comment by singularitti** [comment]
Great! Thanks! I think we should also allow to change the color of left sidebar(line numbers)? I forgot to mention it in the top.

**Comment by singularitti** [comment]
For the third request, maybe we can let xi to detect the system theme and change according to it? I do not know how it is implemented but in some apps they could follow the system's theme.

**Comment by jansol** [comment]
One thing I'd like to see that's kind of related to this is adopting translucent VisualEffectView + vibrancy (as demonstrated at [WWDC](https://developer.apple.com/videos/play/wwdc2018/218)) on *all the things*. (I love using vim in a translucent terminal.)

As for automatically changing between light and dark... xi-mac could allow that iff the current colorscheme has both a light and a dark version.

**Comment by habovh** [comment]
AFAIK and according to [the docs](https://developer.apple.com/documentation/appkit/supporting_dark_mode_in_your_interface) if the app uses semantic colors from `NSColor`, the dark mode support and the transition would be automatically handled by the system.

**Comment by jansol** [comment]
There was a mention about having to use a property instead of setting the color manually in the WWDC video. Setting it manually would mean having to update it manually, whereas the property is supposed to propagete updates from the system automagically.

**Comment by habovh** [comment]
If you comment out [these lines](https://github.com/xi-editor/xi-mac/blob/1fa66c82f52c49137a61a286987451ee28ea66c3/XiEditor/EditViewController.swift#L152-L156), the title bar follows the system theme on first opening window as well. I'm not sure what these lines are here for though, so it definitely needs a more insightful research.

@jansol I think (correct me if I'm wrong) themes are loaded from core and are read as `rgba` colors from a JSON-like object. This would require a "special" theme that does not rely on core, and that uses `NSColor` semantic colors directly. Shouldn't the `NSColor` semantic colors used manually automagically the update as well?

**Comment by singularitti** [comment]
It would be better if we could have the sidebar to be dark also. It is kind of inconsistent in current build:
<img width="728" alt="image" src="https://user-images.githubusercontent.com/25192197/46376701-a447d380-c664-11e8-85a4-b57afd99e1ba.png">

**Comment by singularitti** [comment]
Only commenting out [those lines](https://github.com/xi-editor/xi-mac/blob/1fa66c82f52c49137a61a286987451ee28ea66c3/XiEditor/EditViewController.swift#L152-L156) will result in a weird searchbar:
<img width="614" alt="image" src="https://user-images.githubusercontent.com/25192197/46376819-01dc2000-c665-11e8-8c1d-47c8552f438b.png">

**Comment by habovh** [comment]
@singularitti yes, there's definitely more to it than just a few lines to change...

I've managed to add a *macOS* entry in the Debug → Theme menu, but my approach is rather sloppy.
If I can get something that kind of works as a proof of concept, I might open a PR. Here's a preview.
<img width="730" alt="capture d ecran 2018-10-03 a 00 09 52" src="https://user-images.githubusercontent.com/4186230/46380099-5b624000-c6a1-11e8-8b7d-0bb56a0512cd.png">

**Comment by singularitti** [comment]
It looks pretty good to me! @habovh

**Comment by habovh** [comment]
@singularitti It might look good on a screenshot, but I don't believe the code stands for the high-quality standard that xi-editor is aiming at. This is my first time tinkering with Swift, and like I said, xi-mac has a strong dependency on xi-core, even for the theme appearance. I'll have to dig deeper into this.

**Comment by jansol** [comment]
My main concern is how that interacts with syntax themes. The vibrance blending mode will probably make that magically Just Work, but it'll need some experimenting.

Don't worry about code quality for now, sorting out that is part of the reason why PR reviews exist. WIP/RFC pull requests also have a place, and this sort of thing is exactly that place.

And last but not least, we make a point of not biting in reviews 😉.

**Comment by nangtrongvuon** [comment]
@habovh You can make a PR, and I'll make sure to give it a look :)
